### PR TITLE
feat: RLN proofs as a lightpush service

### PR DESCRIPTION
**Pull request by shash256**
# Description
- Extend lightpush service protocol to attach RLN proofs to messages received from clients (Issue #2593)
- Ensure that the proof is valid. If the service node was able to generate a valid proof, publish the message, otherwise return an error in the lightpush response.

# Changes
- Moved PushHandler to waku_lightpush/callbacks
- It can optionally can take in WakuRLNRelay, and attach RLN proof to messages received.
- If not enabled, the messages can still get published without any additions
- Unit test covering scenario when RLN enabled (and existing e2e test covers the scenario when not enabled)

### CODE REVIEW

File: waku/waku_lightpush/callbacks.nim
Comment (jm-clius) [inline waku/waku_lightpush/callbacks.nim:47]: Preferably these logs, related to mounting, should be in the `waku_node` not in the callbacks module. I'd move the logic of choosing which push handler to "get" to the `waku_node`: if relay is nil, log this case and `getNilPushHandler()`, else `getRelayPushHandler()`.

File: waku/waku_lightpush/callbacks.nim
Comment (jm-clius) [inline waku/waku_lightpush/callbacks.nim:53]: It's probably better to do an early return-on-error here, e.g.:

```
if rlnResultInfo != "":
  return err(rlnResultInfo)
```

File: waku/waku_lightpush/callbacks.nim
Comment (jm-clius) [inline waku/waku_lightpush/callbacks.nim:19]: Return a `Result[WakuMessage, cstring]`

https://status-im.github.io/nim-style-guide/errors.result.html

**Review by jm-clius — CHANGES_REQUESTED** [review]
Thanks! Some comments below, but generally LGTM

**Review by SionoiS — CHANGES_REQUESTED** [review]
The logic is good but code organisation I don't like.

It seams like the `RelayPushHandler` no longer has only 1 purpose. I would split the handler in 2 (proofs & relay), updating the logic in `protocol.nim` in the process.

Also, get familiar with `.isOkOr:` and `.valueOr:` those are great line savers.

**Review by jm-clius — APPROVED** [review]
LGTM. Some code organisation can be done before merging, but no need to re-request review.

@SionoiS it will be good to understand which specific comments are blocking approval? Minor code re-organisation can generally be done without requiring another approval or be left for subsequent PRs. I agree with most of your proposals, but think splitting the attach-RLN-proof and publish-RLN handlers is out of scope here: it would imply, among other things, changing the lightpush protocol to accept multiple push handlers with strong dependency on the order in which these handlers are called. Could be done, but I think it needs a proper design and is best left for when we introduce multiple services to the lightpush service node.

**Review by SionoiS — APPROVED** [review]

File: waku/waku_lightpush/callbacks.nim
Comment (NagyZoltanPeter) [inline waku/waku_lightpush/callbacks.nim:41]: Is it necessary to have proc producing another proc? 
Isn't it possible to define the necessary callback proc here and just use it as handler there?

File: tests/node/test_wakunode_lightpush.nim
Comment (NagyZoltanPeter) [inline tests/node/test_wakunode_lightpush.nim:186]: Can we make a/some negative case here?

**Review by NagyZoltanPeter — APPROVED** [review]
Thank you for this great work, it is an awaited feature I think, at least I need it for lite-protocol-tester to run on RLN-ed networks.
I approve it with some comment for you to consider.

**Pull request merged**
